### ISSUE TITLE
Show input numbers on the Pipeline assembly page

### ISSUE THREAD

**Issue by rhliang** [cfe-lab/Kive#402]
Is there much demand for this?  I just created a trivial testing pipeline that looked like this:
![inputordering](https://cloud.githubusercontent.com/assets/5862525/8627337/35659162-26ff-11e5-9e4b-5a47179f2256.png)

and then got confused why it didn't do what I thought it would do until I saw that the input ordering had changed:
![inputorderingselect](https://cloud.githubusercontent.com/assets/5862525/8627342/427a6a8a-26ff-11e5-8fb3-15e431d4ae9b.png)

I realize that I should have known better here, especially with the isometric grid, but I think this could still be confusing sometimes.

## Summary of the conversation ##
Ordering of inputs affects the form for launching a run, and it should be based on the isometric y-axis position of the input nodes in the diagram.
* [x] Input numbers will be displayed in the input nodes' labels: i1, i2, i3.
* [x] Step numbers will be displayed if the order is ambiguous. They will be displayed in the step nodes' labels: s1, s2, s3.
* [x] Autolayout must not change the input order or the step order. Either adjust the y-axis position enough to make the calculation match the previous order, or include an override value for position that gets cleared when you move an input node.

**Comment by donkirkby** [comment]
I'm not convinced that the numbers will be much clearer than what we already have. I suggest we close this issue and leave it as is until we get more feedback from users.

**Comment by rhliang** [comment]
Several of us at this meeting (sorry Don) have come across this/been persuaded by those of us who have.  However we're going to bump this up slightly in priority because the order was actually thought to be based on the *magnet* order of the methods, which it doesn't seem to be.

**Comment by ArtPoon** [comment]
This should proceed based on method order and then magnet order within methods, which is not based on isometric display.

**Comment by jnak-cfe** [comment]
A clue, maybe? Magnets are sorted by dataset_idx.

drydock_objects.js line 688
```
        // Sort the magnets by index
        var sorted_inputs = this.inputs.sort(function(a,b){ return a.dataset_idx - b.dataset_idx; }),
            sorted_outputs = this.outputs.sort(function(a,b){ return a.dataset_idx - b.dataset_idx; }),
```

**Comment by ArtPoon** [comment]
Have a big label next to each pipeline input indicating its order, like we are presently doing for Methods.

**Comment by rhliang** [comment]
Lines 96-97 of `pipeline_load.js` read, in the function `my.Pipeline.prototype.serialize`:

```
        pipeline_inputs.sort(Geometry.isometricSort);
        pipeline_outputs.sort(Geometry.isometricSort);
```

That indicates to me that the inputs are being saved to the database in straight isometric order, ignoring method order and magnet order.  Should we change this or leave it as is?

**Comment by jnak-cfe** [comment]
I thought, "there's no way"... and then I tried it and you're right. Nice discovery @rhliang.

**Comment by donkirkby** [comment]
The comments on this issue say that we were planning to make the order based on method number and magnet number, but I thought we discussed basing it on isometric order so the pipeline developer could choose the input order. Displaying the input number would remove the current confusion.

**Comment by rhliang** [comment]
Yes that's what I remember too.  I vote that we leave it isometrically ordered.

**Comment by jnak-cfe** [comment]
Prepending a number order to each input could be confusing to the user. Order is only relevant within each method node. As well, we don't want the user to confuse input order with method order. Having multiple lists of ordered nodes might be difficult to keep visually contained.

**Comment by donkirkby** [comment]
Order of inputs is relevant for the form when you launch a run, @jnak-cfe.

You're right, though, that numbering two separate lists is confusing.

Some suggestions, none of which I love:
1. Use A:, B:, C: for inputs, and 1:, 2:, 3: for steps.
2. Use i1:, i2:, i3: for inputs, and s1:, s2:, s3: for steps.
3. Use Roman numerals for inputs.

**Comment by rhliang** [comment]
Of those, I vote for option 2, if only because the thought of input XVIII doesn't appeal to me.

**Comment by jnak-cfe** [comment]
What about putting step numbers in blue circles, and input numbers in red squares? or something like that.
I also want to bring up that this is going to make the autolayout feature produce some ugly results—unless we keep it like it currently is, which reorders inputs into magnet order.

**Comment by rhliang** [comment]
Oh yeah, I didn't think of that.  Unfortunately I think we can't leave it that way, because if we were to hit autolayout, we could change the order of inputs.  Dang.

**Comment by donkirkby** [comment]
Good point about the autolayout, @jnak-cfe. Perhaps we could do the current layout algorithm, then move the inputs along the isometric "x" axis until they have the previous ordering.

As for how to display the input numbers, how about writing the number on the input cylinder or cube? That would make them distinct from the step numbers that appear in the step labels.

![kive_numbered_inputs](https://cloud.githubusercontent.com/assets/1639148/16468596/85e17812-3e01-11e6-8ee7-d0bbdde9269f.png)

Just to be clear, the Roman numerals option was a joke.

**Comment by rhliang** [comment]
I believe that we'd need to adjust the isometric y-axis, not the x-axis, is that right?

I had been thinking I'd try and tackle this issue, but I think this has officially exceeded my JavaScript pay grade.

I think I would have liked to see the Roman numerals.  Maybe someday we can offer interface skins.  Is there a Braille font?

**Comment by jnak-cfe** [comment]
+1 for skinning the interface!

We can definitely adapt the autolayout function, it just might not be as pretty. As for putting labels on the nodes themselves, I'm afraid of that feature becoming too prominent and dominating the canvas with information that is not necessarily important. I think putting them in the labels like the methods indicates that this is still "ordering", but there is also a difference.

**Comment by donkirkby** [comment]
OK, to summarize the conversation so far:
1. Ordering of inputs affects the form for launching a run.
2. Ordering should be based on the isometric y-axis position of the input nodes in the diagram.
3. Input numbers will be displayed in the input nodes' labels: i1, i2, i3.
4. Step numbers will be displayed if the order is ambiguous. They will be displayed in the step nodes' labels: s1, s2, s3.
5. Autolayout must not change the input order. Either adjust the y-axis position enough to make the calculation match the previous order, or include an override value for position that gets cleared when you move an input node.

**Comment by jnak-cfe** [comment]
Sounds good, @donkirkby

**Comment by rhliang** [comment]
That all looks good.  Do we want to add anything for output nodes?  They aren't important right now except for navigability of the View Results page.  Someday if we start doing nested Pipelines in the interface, it could become more important.

**Comment by rhliang** [comment]
I think I can maybe tackle this JavaScript again.  I'm going to have a stab at it.

**Comment by donkirkby** [comment]
Let's not worry about outputs yet, @rhliang. When you look at input ordering during autolayout, compare how it handles the ordering of steps. Those seem to keep the same ordering.

**Comment by rhliang** [comment]
Oh oops I didn't mean I was going to look at autolayout just yet.  I was just going to add i1, i2, i3 to the input labels for now.

**Comment by donkirkby** [comment]
I've added a summary and task list to the issue description. Finish whichever tasks you like, @rhliang, then unassign yourself. If we decide to bump the autolayout to the next release, we can split the issue in two.

**Comment by rhliang** [comment]
I made a band-aid fix (commit 79bd27ee16a925b9239e8ef7e6751be480da5910), and talked with James about a better solution to the problem.  It would take a little deeper knowledge than I have at the moment.

**Comment by jnak-cfe** [comment]
Autolayout issue was solved by re-positioning all input nodes at the firstmost iso-y-position and calculating their iso-x-position based solely on their stated order.

A bit different than discussed, but much, much easier to implement.
